src/azul/openapi/validation.py

```diff
@@ -0,0 +1,242 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from copy import (
+    copy,
+)
+import json
+import logging
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    TypeVar,
+)
+
+import attr
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    RequirementError,
+)
+from azul.openapi.schema import (
+    python_type_for,
+)
+from azul.plugins.metadata.hca.transform import (
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, kw_only=False)
+class ValidationError(Exception):
+    message: str
+    name: str
+    value: Any
+    schema: Optional[JSON] = None
+    content: Optional[JSON] = None
+
+    def to_dict(self):
+        """
+        :return: Dictionary of atttributes that are not None
+        """
+        return {
+            attribute.name: self.__getattribute__(attribute.name)
+            for attribute in self.__attrs_attrs__
+            if self.__getattribute__(attribute.name) is not None
+        }
+
+
+V = TypeVar('V', bound='SpecValidator')
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class SpecValidator(ABC):
+    """
+    A base class for OpenAPI specification validators. See
+    https://swagger.io/specification/#parameter-object for details about
+    parameter specification. The sub-classes for this class are broken down
+    by the complexity of the parameter that is being defined.
+    """
+    name: str
+    required: bool
+    in_: str
+    description: Optional[str] = None
+    deprecated: bool = False
+    default: Optional[str] = None
+    schema: Optional[JSON] = None
+
+    @abstractmethod
+    def validate(self, param_value: str):
+        """
+        Validates if a given request parameter meets the specification.
+        :param param_value: The value of the parameter to be validated
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def if_missing(self):
+        """
+        Used to return information about parameter if it is missing from a
+        request
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def from_spec(cls, spec: JSON) -> V:
+        """
+        Returns the applicable specification validator for a parameter
+        """
+        # noinspection PyTypeChecker
+        kwargs: Dict[str, JSON] = copy(spec)
+        kwargs['in_'] = kwargs.pop('in')
+        if 'content' in kwargs:
+            assert 'schema' not in kwargs
+            return ContentSpecValidator(**kwargs)
+        elif 'schema' in kwargs:
+            return SchemaSpecValidator(**kwargs)
+        else:
+            assert False
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class SchemaSpecValidator(SpecValidator):
+    """
+    SchemaSpecValidators are used to validate parameters that describe
```

Comment:
Parameters don't describe schemas. The other way around, maybe. This problem seems to be recurring in the other docstrings.